File: sidecar/src/main/java/no/horten/modcut/DriverCatalog.java
package no.horten.modcut;

import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.node.ArrayNode;
import com.fasterxml.jackson.databind.node.ObjectNode;
import de.thomas_oster.liblasercut.LaserCutter;
import de.thomas_oster.liblasercut.drivers.EpilogHelix;
import de.thomas_oster.liblasercut.drivers.EpilogZing;
import java.util.List;

/** Single authoritative sidecar allow-list. Only entries with a complete execution path belong here. */
final class DriverCatalog {
  record Descriptor(String id, String displayName, String manufacturer, String family, String protocol,
      List<String> connectionTypes, Integer defaultPort, String fileExtension, List<Integer> rasterDpis,
      int vectorDpi, boolean nativeRaster, boolean softwareFocus, boolean controlledZ, boolean framing,
      boolean softwareCancel, double focusMinMm, double focusMaxMm) {
    boolean epilog() { return id.startsWith("epilog-"); }
    LaserCutter createEpilog(String host, int port, double bedWidth, double bedHeight) {
      LaserCutter cutter = switch (id) {
        case "epilog-zing" -> new EpilogZing(host);
        case "epilog-helix" -> new EpilogHelix(host);
        default -> throw new IllegalStateException("Ikke en Epilog-driver: " + id);
      };
      if (cutter instanceof EpilogZing value) { value.setPort(port); value.setBedWidth(bedWidth); value.setBedHeight(bedHeight); value.setAutoFocus(false); value.setHideSoftwareFocus(false); }
      if (cutter instanceof EpilogHelix value) { value.setPort(port); value.setBedWidth(bedWidth); value.setBedHeight(bedHeight); value.setAutoFocus(false); value.setHideSoftwareFocus(false); }
      return cutter;
    }
  }
  static final Descriptor DUMMY = new Descriptor("dummy", "Dummy", "Generic", "Simulation", "none", List.of("usb"), null, ".gcode", List.of(), 300, false, false, true, true, true, 0, 0);
  static final Descriptor GRBL = new Descriptor("grbl", "GRBL", "Generic", "GRBL", "GRBL serial/TCP", List.of("usb", "network"), 23, ".gcode", List.of(), 300, false, false, true, true, true, 0, 0);
  static final Descriptor ZING = new Descriptor("epilog-zing", "Epilog Zing", "Epilog", "Zing", "LPD/PJL/PCL/HPGL", List.of("network"), 515, ".prn", List.of(100,200,250,400,500,1000), 500, true, true, false, true, false, -12.6, 12.6);
  static final Descriptor HELIX = new Descriptor("epilog-helix", "Epilog Helix", "Epilog", "Helix", "LPD/PJL/PCL/HPGL", List.of("network"), 515, ".prn", List.of(75,150,200,300,400,600,1200), 600, true, true, false, true, false, -12.6, 12.6);
  static final List<Descriptor> ALL = List.of(DUMMY, GRBL, ZING, HELIX);
  static Descriptor resolve(String value) {
    String key = value == null ? "" : value.trim().toLowerCase();
    key = switch (key) { case "dummy" -> "dummy"; case "grbl" -> "grbl"; case "epilog zing" -> "epilog-zing"; case "epilog helix" -> "epilog-helix"; default -> key; };
    String id = key;
    return ALL.stream().filter(d -> d.id().equals(id)).findFirst().orElseThrow(() -> new IllegalArgumentException("Driveren støttes ikke: " + value + "."));
  }
  static ObjectNode json(ObjectMapper mapper) {
    ObjectNode out = mapper.createObjectNode(); ArrayNode drivers = out.putArray("drivers");
    for (Descriptor d : ALL) { ObjectNode n=drivers.addObject(); n.put("id",d.id()); if (d.id().equals("grbl")) n.put("legacyName", "Grbl"); n.put("displayName",d.displayName()); n.put("manufacturer",d.manufacturer()); n.put("family",d.family()); n.put("protocol",d.protocol()); n.putPOJO("connectionTypes",d.connectionTypes()); if(d.defaultPort()!=null)n.put("defaultPort",d.defaultPort()); n.put("fileExtension",d.fileExtension()); n.putPOJO("rasterDpis",d.rasterDpis()); n.put("vectorDpi",d.vectorDpi()); n.put("nativeRaster",d.nativeRaster()); n.put("softwareFocus",d.softwareFocus()); n.put("controlledZ",d.controlledZ()); n.put("framing",d.framing()); n.put("softwareCancel",d.softwareCancel()); }
    out.put("library","LibLaserCut"); return out;
  }
  private DriverCatalog() {}
}

File: renderer/driver-catalog.mjs
export const DRIVER_CATALOG = Object.freeze([
  { id: "dummy", displayName: "Dummy", manufacturer: "Generic", family: "Simulation", protocol: "none", connectionTypes: ["usb"], defaultPort: null, fileExtension: ".gcode", rasterDpis: [], vectorDpi: 300, nativeRaster: false, softwareFocus: false, controlledZ: true, framing: true, softwareCancel: true, available: true, defaults: { maxFeed: 12000 } },
  { id: "grbl", displayName: "GRBL", manufacturer: "Generic", family: "GRBL", protocol: "GRBL serial/TCP", connectionTypes: ["usb", "network"], defaultPort: 23, fileExtension: ".gcode", rasterDpis: [], vectorDpi: 300, nativeRaster: false, softwareFocus: false, controlledZ: true, framing: true, softwareCancel: true, available: true, defaults: { baud: 115200, maxFeed: 12000 } },
  { id: "epilog-zing", displayName: "Epilog Zing", manufacturer: "Epilog", family: "Zing", protocol: "LPD/PJL/PCL/HPGL", connectionTypes: ["network"], defaultPort: 515, fileExtension: ".prn", rasterDpis: [100, 200, 250, 400, 500, 1000], vectorDpi: 500, nativeRaster: true, softwareFocus: true, focusMinMm: -12.6, focusMaxMm: 12.6, controlledZ: false, framing: true, softwareCancel: false, available: true, defaults: { maxFeed: 12000 } },
  { id: "epilog-helix", displayName: "Epilog Helix", manufacturer: "Epilog", family: "Helix", protocol: "LPD/PJL/PCL/HPGL", connectionTypes: ["network"], defaultPort: 515, fileExtension: ".prn", rasterDpis: [75, 150, 200, 300, 400, 600, 1200], vectorDpi: 600, nativeRaster: true, softwareFocus: true, focusMinMm: -12.6, focusMaxMm: 12.6, controlledZ: false, framing: true, softwareCancel: false, available: true, defaults: { maxFeed: 12000 } },
]);
export const MACHINE_PRESETS = Object.freeze([
  { id: "generic-dummy", manufacturer: "Generic", model: "Dummy", driverId: "dummy", name: "Dummy (offline)", bedW: 600, bedH: 400 },
  { id: "generic-grbl", manufacturer: "Generic", model: "GRBL", driverId: "grbl", name: "GRBL laser", bedW: 600, bedH: 400 },
  { id: "epilog-zing", manufacturer: "Epilog", model: "Zing", driverId: "epilog-zing", name: "Epilog Zing", bedW: 600, bedH: 300 },
  { id: "epilog-helix", manufacturer: "Epilog", model: "Helix", driverId: "epilog-helix", name: "Epilog Helix", bedW: null, bedH: null },
]);
const LEGACY_IDS = new Map([["dummy", "dummy"], ["grbl", "grbl"], ["epilog zing", "epilog-zing"], ["epilog helix", "epilog-helix"]]);
export const driverById = (id) => DRIVER_CATALOG.find((driver) => driver.id === id);
export const normalizeDriverId = (value) => LEGACY_IDS.get(String(value || "").trim().toLowerCase()) || String(value || "").trim().toLowerCase();
export function normalizeMachineProfile(machine) {
  const normalized = structuredClone(machine);
  normalized.driverId = normalizeDriverId(normalized.driverId || normalized.driver);
  const driver = driverById(normalized.driverId) || DRIVER_CATALOG[0];
  normalized.driverId = driver.id; normalized.driver = driver.id;
  normalized.manufacturer ||= driver.manufacturer; normalized.model ||= driver.family;
  normalized.conn ||= {};
  if (!driver.connectionTypes.includes(normalized.conn.type)) normalized.conn.type = driver.connectionTypes[0];
  if (normalized.conn.type === "network" && (!normalized.conn.port || (normalized.conn.port === 23 && driver.defaultPort === 515))) normalized.conn.port = driver.defaultPort;
  return normalized;
}
export const capabilitiesForMachine = (machine) => driverById(normalizeDriverId(machine?.driverId || machine?.driver)) || DRIVER_CATALOG[0];
